modules/mono/glue/GodotSharp/GodotSharp/Core/Vector4.cs

```diff
@@ -0,0 +1,961 @@
+        /// </summary>
+        /// <param name="onNormal">The vector to project onto.</param>
+        /// <returns>The projected vector.</returns>
+        public Vector4 Project(Vector4 onNormal)
+        {
+            return onNormal * (Dot(onNormal) / onNormal.LengthSquared());
+        }
+
+        /// <summary>
+        /// Returns this vector reflected from a plane defined by the given <paramref name="normal"/>.
+        /// </summary>
+        /// <param name="normal">The normal vector defining the plane to reflect from. Must be normalized.</param>
+        /// <returns>The reflected vector.</returns>
+        public Vector4 Reflect(Vector4 normal)
+        {
+#if DEBUG
+            if (!normal.IsNormalized())
+            {
+                throw new ArgumentException("Argument is not normalized", nameof(normal));
+            }
+#endif
+            return (2.0f * Dot(normal) * normal) - this;
+        }
+
+        /// <summary>
+        /// Returns this vector with all components rounded to the nearest integer,
+        /// with halfway cases rounded towards the nearest multiple of two.
+        /// </summary>
+        /// <returns>The rounded vector.</returns>
+        public Vector4 Round()
+        {
+            return new Vector4(Mathf.Round(x), Mathf.Round(y), Mathf.Round(z), Mathf.Round(w));
+        }
+
+        /// <summary>
+        /// Returns a vector with each component set to one or negative one, depending
+        /// on the signs of this vector's components, or zero if the component is zero,
+        /// by calling <see cref="Mathf.Sign(real_t)"/> on each component.
+        /// </summary>
+        /// <returns>A vector with all components as either <c>1</c>, <c>-1</c>, or <c>0</c>.</returns>
+        public Vector4 Sign()
+        {
+            Vector4 v;
+            v.x = Mathf.Sign(x);
+            v.y = Mathf.Sign(y);
+            v.z = Mathf.Sign(z);
+            v.w = Mathf.Sign(w);
+            return v;
+        }
+
+        /// <summary>
+        /// Returns this vector with each component snapped to the nearest multiple of <paramref name="step"/>.
+        /// This can also be used to round to an arbitrary number of decimals.
+        /// </summary>
+        /// <param name="step">A vector value representing the step size to snap to.</param>
+        /// <returns>The snapped vector.</returns>
+        public Vector4 Snapped(Vector4 step)
+        {
+            return new Vector4
+            (
+                Mathf.Snapped(x, step.x),
+                Mathf.Snapped(y, step.y),
+                Mathf.Snapped(z, step.z),
+                Mathf.Snapped(w, step.w)
+            );
+        }
+
+        // Constants
+        private static readonly Vector4 _zero = new Vector4(0, 0, 0, 0);
+        private static readonly Vector4 _one = new Vector4(1, 1, 1, 1);
+        private static readonly Vector4 _inf = new Vector4(Mathf.Inf, Mathf.Inf, Mathf.Inf, Mathf.Inf);
```

Comment:
It's likely worth adding some constants for the basic unit vectors. It's not sensible to give friendly names like "up" and "right", but we can still declare some constants.

```cs
        private static readonly Vector4 _unitX = new Vector4(1, 0, 0, 0);
        private static readonly Vector4 _unitY = new Vector4(0, 1, 0, 0);
        private static readonly Vector4 _unitZ = new Vector4(0, 0, 1, 0);
        private static readonly Vector4 _unitW = new Vector4(0, 0, 0, 1);
```